### PR TITLE
Add spec to sync Makefile rules with compose

### PR DESCRIPTION
**Pull request by benthorner**
https://github.com/alphagov/govuk-docker/issues/237

Previously we relied on people manually listing Makefile prerequisites to
match the dependencies in the compose file for that service. This has lead
to some Makefiles being out-of-sync. This adds a spec to help bring the
Makefiles back in sync with compose for app dependencies.

### CODE REVIEW

File: spec/make_spec.rb [hunk 1]
```diff
@@ -29,8 +29,8 @@ def compose_strip_stacks(dependencies)
   end
 
   def makeable_app_services
-    @app_services ||= service_names.select do |service_name|
-      File.exists?(make_file(service_name)) &&
+    @makeable_app_services ||= service_names.select do |service_name|
+      File.exist?(make_file(service_name)) &&
```
Comment (ChrisBAshton) [inline spec/make_spec.rb]: This and line 32 should be fixed up to the previous commit.

File: spec/make_spec.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      expect(compose_dependencies(service_name))
+        .to match_array(makefile_dependencies(service_name))
+    end
+  end
+
+  def compose_dependencies(service_name)
+    filename = compose_file(service_name)
+    service_stacks = YAML.load_file(filename)["services"].values
+    dependencies = service_stacks.flat_map { |s| s["depends_on"].to_a }
+    dependencies = compose_strip_stacks(dependencies).uniq
```
[the remark this answers is not in the document]
Comment (benthorner) [reply spec/make_spec.rb]: Oh yeah. Logically it seemed neater to to it at the point of origin, but I agree it's not necessary. Removed.

File: spec/make_spec.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  def makefile_dependencies(service_name)
+    filename = make_file(service_name)
+    return [] unless File.exists?(filename)
+
+    targets = File.readlines(filename).first.scan(/[\w\-_]+/)
+    (targets & makeable_app_services) - [service_name]
+  end
+
+  def compose_strip_stacks(dependencies)
```
[the remark this answers is not in the document]
Comment (benthorner) [reply spec/make_spec.rb]: Sure - done.

File: spec/make_spec.rb [hunk 4]
```diff
@@ -18,7 +18,7 @@ def compose_dependencies(service_name)
 
   def makefile_dependencies(service_name)
     filename = make_file(service_name)
-    return [] unless File.exists?(filename)
+    return [] unless File.exist?(filename)
```
[the remark this answers is not in the document]
Comment (benthorner) [reply spec/make_spec.rb]: Ah, good spot - fixed.

**Comment by benthorner** [comment]
Thanks for reviewing @ChrisBAshton :-). I think this is ready for another look.

**Pull request merged**